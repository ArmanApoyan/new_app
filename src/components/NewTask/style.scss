.newTask {
  display: flex;
  flex-direction: column;
  width: 80%;
  gap: 15px;
  margin: 10px auto;
  background-color: white;
  padding: 20px;
  border-radius: 10px;
  input,
  select,
  textarea {
    height: 30px;
    font-size: 18px;
    border: 2px solid lightgray;
    border-radius: 10px;
  }
  textarea {
    height: 150px;
    resize: none;
    &:focus {
      font-size: 16px;
    }
  }
  button {
    background-color: rgb(51, 141, 129);
    color: white;
    font-weight: bold;
    width: 120px;
    height: 30px;
    margin: 10px auto;
    border: 2px solid lightgray;
    border-radius: 10px;
    font-size: 18px;
    &:hover {
      cursor: pointer;
      border: 2px solid darkgreen;
    }
  }
}
.redborder {
  border: 2px solid red;
}
.error {
  color: red;
  height: 0px;
  
}
.view {
  margin: 15px;
  p {
    font-size: 25px;
    padding: 5px;
    border-radius: 10px;
    cursor: default;
    &:hover {
      color: darkblue;
    }
  }
  .status {
    border: none;
    font-weight: bold;
    cursor: default;
    &:hover {
      border: none;
      color: black;
    }
  }
}

@media (max-width: 600px) {
  .newTask {
    select,
    input,
    textarea {
      font-size: 15px;
    }
  }
  .view {
    p {
      font-size: 22px;
    }
  }
}
@media (max-width: 400px) {
  .newTask {
    select,
    input,
    textarea {
      font-size: 13px;
    }
  }
  .view {
    p {
      font-size: 20px;
    }
  }
}
