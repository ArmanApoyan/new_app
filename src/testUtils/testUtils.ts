import { combineReducers, createStore } from "redux";
import { myReducer } from "../store/Task/reducer";


export function createTestStore() {
    const store = createStore(
      combineReducers({
        task: myReducer,
      })
    );
    return store;
  }