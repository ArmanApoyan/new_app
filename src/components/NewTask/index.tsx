import { getIds, handleBlur, handleChange, handleDblClick, random, validator } from "../../utils";
import React, { useEffect, useState, useMemo, memo } from "react";
import { Column, Goal, State } from "../../types/global";
import { useDispatch, useSelector } from "react-redux";
import { action1 } from "../../store/Task/action";
import { BiArrowBack } from "react-icons/bi";
import "./style.scss";

interface Props {
  close: CallableFunction;
  type: string;
  task?: Goal;
}

const NewTask: React.FC<Props> = (props) => {
  const { close, type, task } = props;
  const dispatch = useDispatch();

  const [changeType, setChangeType] = useState("");
  const [compType, setcompType] = useState(type);

  const { columns, goals } = useSelector((state: State) => state.task);
  const newIds = useMemo(getIds(goals), [goals]);
  const [formData, setFormData] = useState({
    id: random(newIds),
    title: { value: "", error: false },
    description: { value: "", error: false },
    status: { value: "", error: false },
  });

  useEffect(() => {
    if (task) {
      setFormData({
        id: task.id,
        title: { value: task.title, error: false },
        description: { value: task.description, error: false },
        status: { value: task.status, error: false },
      });
    }
  }, [task]);

  const clickBack = () =>{
    setcompType("view");
    if (task) {
      setFormData({
        id: task.id,
        title: { value: task.title, error: false },
        description: { value: task.description, error: false },
        status: { value: task.status, error: false },
        })}
  }
  const handleSubmit = (e: any) => {
    e.preventDefault();
    if (formData.status.value == "default" || formData.status.value == "") {
      setFormData({ ...formData });
      formData.status.error = true;
      return;
    }
    if (formData.title.value == "") {
      formData.title.error = true;
      setFormData({ ...formData });
      return;
    }
    if (formData.title.error == true) {
      return;
    }
    let data = {
      id: formData.id,
      title: formData.title.value,
      description: formData.description.value,
      status: formData.status.value,
    };
    // @ts-ignore
    dispatch(action1(compType, data));
    close();
  };
  return (
    <>{compType != "view" && (
        <form
          className="newTask"
          onSubmit={(e) => {
            handleSubmit(e);
          }}>
          {formData.status.error == true && (
            <p className="error">Select Column</p>
          )}
          {compType == "add" && (
            <select
              name="status"
              onBlur={(e) => handleBlur(e, formData.status, "default", setFormData, formData)}
              defaultValue={"default"}
              onChange={(e) => {handleChange(e,setFormData,formData)}}>
              <option disabled value="default">
                Select Column
              </option>
              {columns.map((el: Column, i: number) => {
                return (
                  <option value={el.title} key={i}>
                    {el.title}
                  </option>
                );
              })}
            </select>
          )}
          {compType == "create" && changeType == "" && (
            <select
              name="status"
              defaultValue={formData.status.value}
              onChange={(e) => {handleChange(e, setFormData, formData)}}>
              {columns.map((el: Column, i: number) => {
                if (el.title == formData.status.value) {
                  return (
                    <option selected value={el.title} key={i}>
                      {el.title}
                    </option>
                  );
                }
                return (
                  <option value={el.title} key={i}>
                    {el.title}
                  </option>
                );
              })}
            </select>
          )}
          {changeType == "select" && (
            <><select
                name="status"
                defaultValue={formData.status.value}
                onChange={(e) => {handleChange(e, setFormData, formData)}}>
                {columns.map((el: Column, i: number) => {
                  return (
                    <option value={el.title} key={i}>
                      {el.title}
                    </option>
                  );
                })}
              </select>
              <p
                onDoubleClick={() => setChangeType("input")}
                className="view_title title">
                {formData.title.value}
              </p>
              <p
                onDoubleClick={() => setChangeType("textarea")}
                className="view_description des">
                {formData.description.value}
              </p></>
          )}
          {formData.title.error == true && (
            <p className="error">Title is incorrect</p>
          )}
          {changeType == "" && (
            <input
              onBlur={(e) => {
                handleBlur(e, formData.title, "", setFormData, formData);
                if (!validator(/^\D[a-zA-Z0-9,.!? ]{2,}/, e.target.value)) {
                  formData.title.error = true;
                  setFormData({ ...formData });
                }
              }}
              type="text"
              name="title"
              placeholder="Title"
              value={formData.title.value}
              onChange={(e) => {handleChange(e, setFormData, formData)}}/>
          )}
          {changeType == "input" && (
            <><p
                onDoubleClick={() => setChangeType("select")}
                className="view_status">
                {formData.status.value}
              </p>
              <input
                type="text"
                placeholder="Title"
                value={formData.title.value}
                name="title"
                onChange={(e) => {handleChange(e, setFormData, formData)}}/>
              <p
                onDoubleClick={() =>setChangeType("textarea")}
                className="view_description des">
                {formData.description.value}
              </p></>
          )}
          {changeType == "" && (
            <textarea
              name="description"
              onChange={(e) => {handleChange(e, setFormData, formData)}}
              value={formData.description.value}
              placeholder="Description"/>
          )}
          {changeType == "textarea" && (
            <><p
                onDoubleClick={() =>setChangeType("select")}
                className="view_status">
                {formData.status.value}
              </p>
              <p
                onDoubleClick={() =>setChangeType("input")}
                className="view_title title">
                {formData.title.value}
              </p>
              <textarea
              name="description"
              onChange={(e) => {handleChange(e, setFormData, formData)}}
                value={formData.description.value}
                placeholder="Description"/></>)}
          {compType == "add" && (
            <button className="submit" type="submit">
              ADD
            </button>)}
          {compType == "create" && (
            <button className="submit" type="submit">
              Save
            </button>)}
          {compType == "create" && changeType != "" && (
            <button
              className="back"
              onClick={() => clickBack()}>
              <BiArrowBack />
            </button>)}
        </form>
      )}
      {compType == "view" && (
        <div className="view">
            <p className="id">{task?.id}</p>
          <p
            className="status"
            onDoubleClick={() =>handleDblClick("select",setcompType,setChangeType)}>
            {formData.status.value}
          </p>
          <p
            className="title"
            onDoubleClick={() =>handleDblClick("input",setcompType,setChangeType)}>
            {formData.title.value}
          </p>
          <p
            className="des"
            onDoubleClick={() =>handleDblClick("textarea",setcompType,setChangeType)}>
            {formData.description.value}
          </p>
        </div>)}
      </>)}
export default memo(NewTask);