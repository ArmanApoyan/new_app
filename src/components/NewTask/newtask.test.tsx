import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Provider } from "react-redux";
import { createTestStore } from "../../testUtils/testUtils";
import NewTask from "./";

describe("newtask component test", () => {
  it("newtask rendering", () => {
    let store = createTestStore();
    const close = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <NewTask close={close} type="" />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
  it("closing after adding", () => {
    let store = createTestStore();
    const close = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <NewTask close={close} type="add" />
      </Provider>
    );
    const btn = screen.getByText("ADD")
    fireEvent.click(btn)
    expect(container).toMatchSnapshot()
  });
  it("closing after saving", () => {
    let store = createTestStore();
    const close = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <NewTask close={close} type="create" />
      </Provider>
    );
    const btn = screen.getByText("Save")
    fireEvent.click(btn)
    expect(container).toMatchSnapshot()
  });
  it("input changing", () => {
    let store = createTestStore();
    const close = jest.fn();
    render(
      <Provider store={store}>
        <NewTask close={close} type="add" />
      </Provider>
    );
    const inp = screen.getByTestId("inp")
    userEvent.type(inp,"title")
    expect(inp).toHaveValue("title")
  });
  it("textarea changing", () => {
    let store = createTestStore();
    const close = jest.fn();
    render(
      <Provider store={store}>
        <NewTask close={close} type="add" />
      </Provider>
    );
    const textarea = screen.getByTestId("textarea")
    userEvent.type(textarea,"description")
    expect(textarea).toHaveValue("description")
  });
  it("view mode", () => {
    let store = createTestStore();
    const close = jest.fn();
    render(
      <Provider store={store}>
        <NewTask close={close} type="view" />
      </Provider>
    );
    const ps = screen.getAllByTestId("p")
    expect(ps).toMatchSnapshot()
  });
  it("select changing", () => {
    let store = createTestStore();
    const close = jest.fn();
    render(
      <Provider store={store}>
        <NewTask close={close} type="add" />
      </Provider>
    );
    const select = screen.getByTestId("select")
    userEvent.selectOptions(select,"Done")
    expect(select).toHaveValue("Done")
    userEvent.selectOptions(select,"On Progress")
    expect(select).toHaveValue("On Progress")
  });
  it("back button", () => {
    let store = createTestStore();
    const close = jest.fn();
    render(
      <Provider store={store}>
        <NewTask close={close} type="view" />
      </Provider>
    );
    const ps = screen.getAllByTestId("p")
    const value = ps[2].innerText
    fireEvent.dblClick(ps[2])
    const status = screen.getByTestId("inp")
    userEvent.type(status,"asd")
    fireEvent.click(screen.getByTestId("back"))
    expect(screen.getAllByTestId("p")[2].innerText).toEqual(value)
  });
});
