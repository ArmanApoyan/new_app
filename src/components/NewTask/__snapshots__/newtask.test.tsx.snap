// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`newtask component test closing after adding 1`] = `
<div>
  <form
    class="newTask"
  >
    <p
      class="error"
    >
      Select Column
    </p>
    <select
      data-testid="select"
      name="status"
    >
      <option
        disabled=""
        selected=""
        value="default"
      >
        Select Column
      </option>
      <option
        value="To Do"
      >
        To Do
      </option>
      <option
        value="On Progress"
      >
        On Progress
      </option>
      <option
        value="Done"
      >
        Done
      </option>
    </select>
    <input
      data-testid="inp"
      name="title"
      placeholder="Title"
      type="text"
      value=""
    />
    <textarea
      data-testid="textarea"
      name="description"
      placeholder="Description"
    />
    <button
      class="submit"
      type="submit"
    >
      ADD
    </button>
  </form>
</div>
`;

exports[`newtask component test closing after saving 1`] = `
<div>
  <form
    class="newTask"
  >
    <p
      class="error"
    >
      Select Column
    </p>
    <select
      name="status"
    >
      <option
        value="To Do"
      >
        To Do
      </option>
      <option
        value="On Progress"
      >
        On Progress
      </option>
      <option
        value="Done"
      >
        Done
      </option>
    </select>
    <input
      data-testid="inp"
      name="title"
      placeholder="Title"
      type="text"
      value=""
    />
    <textarea
      data-testid="textarea"
      name="description"
      placeholder="Description"
    />
    <button
      class="submit"
      type="submit"
    >
      Save
    </button>
  </form>
</div>
`;

exports[`newtask component test newtask rendering 1`] = `
<div>
  <form
    class="newTask"
  >
    <input
      data-testid="inp"
      name="title"
      placeholder="Title"
      type="text"
      value=""
    />
    <textarea
      data-testid="textarea"
      name="description"
      placeholder="Description"
    />
  </form>
</div>
`;

exports[`newtask component test view mode 1`] = `
Array [
  <p
    class="id"
    data-testid="p"
  />,
  <p
    class="status"
    data-testid="p"
  />,
  <p
    class="title"
    data-testid="p"
  />,
  <p
    class="des"
    data-testid="p"
  />,
]
`;
